mcmc_utils.py

```diff
@@ -263,48 +288,79 @@ def run_ptmcmc_save(sampler, startPos, nSteps, file, progress=True, **kwargs):
             thisPos = zpos[k]
             thisProb = zprob[k]
-            f.write("{0:4d} {1:s} {2:f}\n".format(k, " ".join(map(str, thisPos)), thisProb))
-        f.close()
+
+            with open(file, 'a') as f:
+                f.write("{0:4d} {1:s} {2:f}\n".format(k, " ".join(
+                map(str, thisPos)), thisProb))
+
     if progress:
         bar.close()
     return sampler
 
 
-def flatchain(chain, npars, nskip=0, thin=1):
+def flatchain(chain, npars=None, nskip=0, thin=1):
     '''flattens a chain (i.e collects results from all walkers),
     with options to skip the first nskip parameters, and thin the chain
     by only retrieving a point every thin steps - thinning can be useful when
     the steps of the chain are highly correlated'''
+    if npars is None:
+        npars = chain.shape[2]
     return chain[:, nskip::thin, :].reshape((-1, npars))
 
 
-def readchain(file, nskip=0, thin=1):
-    data = pd.read_csv(file, header=None, compression=None, delim_whitespace=True)
+def readchain(file, **kwargs):
+    '''Reads in the chain file in a single thread.
+    Returns the chain in the shape (nwalkers, nprod, npars)
+    '''
+    data = pd.read_csv(file, header=None, compression=None,
+                       delim_whitespace=True, **kwargs)
     data = np.array(data)
-    nwalkers = int(data[:, 0].max()+1)
-    nprod = int(data.shape[0]/nwalkers)
-    npars = data.shape[1] - 1  # first is walker ID, last is ln_prob
-    chain = np.reshape(data[:, 1:], (nwalkers, nprod, npars))
+
+    # Figure out what shape the result should have.
+    nwalkers = int(np.amax(data[:, 0])+1)
+    nprod = int(data.shape[0] / nwalkers)
+    npars = int(data.shape[1] - 1)
+
+    # empty array to fill. Make it nans to be safe?
+    chain = np.zeros((nwalkers, nprod, npars))
+    chain[:, :, :] = np.nan
+
+    for i in range(nwalkers):
+        chain[i] = data[i::nwalkers, 1:]
+
     return chain
 
 
-def readchain_dask(file, nskip=0, thin=1):
-    data = dd.io.read_csv(file, engine='c', header=None, compression=None,
-                          na_filter=False, delim_whitespace=True)
+def readchain_dask(file, **kwargs):
+    '''Reads in the chain file using threading.
+    Returns the chain in the shape (nwalkers, nprod, npars).'''
+    data = dd.io.read_csv(file, engine='c', header=1, compression=None,
+                          na_filter=False, delim_whitespace=True, **kwargs)
     data = data.compute()
     data = np.array(data)
-    nwalkers = int(data[:, 0].max()+1)
-    nprod = int(data.shape[0]/nwalkers)
-    npars = data.shape[1] - 1  # first is walker ID, last is ln_prob
-    chain = np.reshape(data[:, 1:], (nwalkers, nprod, npars))
+
+    # Figure out what shape the result should have.
+    nwalkers = int(np.amax(data[:, 0])+1)
+    nprod = int(data.shape[0] / nwalkers)
+    npars = int(data.shape[1] - 1)
+
+    # empty array to fill. Make it nans to be safe?
+    chain = np.zeros((nwalkers, nprod, npars))
```

Comment:
Does this still run lazily? The idea here was to parallelise the reading of large chain files to speed things up...

Reply:
Sorry, I'm not sure what you mean by 'lazily'... It's still threaded if that's what you mean. The dd.io.read_csv() is the threaded part, the rest is just massaging into the shape (nwalkers, nprod, npars). 

I changed this because the old version of the code doesn't actually collect the walkers properly. They're stored like
```
walker 0, step 0
walker 1, step 0
...
walker N, step 0
walker 0, step 1
walker 1, step 1
...
walker N, step 1
```

etc., but the way it was being handled returned the first `nprod` lines of the file as all the walker 0 steps, then the next `nprod` lines as the steps of walker 1, and so on. This fixes that.

Reply:
I see - lazy means that nothing is actually done. Dask works by collecting operations together into a graph of things to be done, that is then evaluated on as many parallel nodes as possible. dd.io.read_csv command actually does nothing, its the ```dd.compute``` line that actually reads in the data. So whatever you do after the ```compute``` line works with an actual, in memory array.